crypto/txscript/src/lib.rs

```diff
@@ -973,6 +1047,231 @@ mod tests {
+            let hash = calc_ecdsa_signature_hash(&tx.as_verifiable(), 0, SIG_HASH_ALL, reused);
+            let msg = secp256k1::Message::from_digest_slice(hash.as_bytes().as_slice()).unwrap();
+            let sig = keypair.secret_key().sign_ecdsa(msg);
+            let mut signature = sig.serialize_compact().to_vec();
+            signature.push(SIG_HASH_ALL.to_u8());
+            SignatureData { signature, public_key: keypair.public_key().serialize().to_vec() }
+        };
+
+        let test_cases = vec![
+            // Basic Schnorr CheckSig
+            TestCase {
+                name: "Basic Schnorr CheckSig",
+                script_builder: Box::new(|sb| sb.add_op(OpCheckSig)),
+                sig_builder: Box::new(move |tx, reused| SignatureScriptBuilder::Single(create_schnorr_signature(tx, reused))),
+                expected_sig_ops: 1,
+                sig_op_limit: 1,
+                should_pass: true,
+            },
+            // Basic ECDSA CheckSig
+            TestCase {
+                name: "Basic ECDSA CheckSig",
+                script_builder: Box::new(|sb| sb.add_op(OpCheckSigECDSA)),
+                sig_builder: Box::new(move |tx, reused| SignatureScriptBuilder::Single(create_ecdsa_signature(tx, reused))),
+                expected_sig_ops: 1,
+                sig_op_limit: 1,
+                should_pass: true,
+            },
+            // Mixed Schnorr and ECDSA
+            TestCase {
+                name: "Mixed Schnorr and ECDSA",
+                script_builder: Box::new(|sb| sb.add_op(OpCheckSigVerify)?.add_op(OpCheckSigECDSA)),
+                sig_builder: Box::new(move |tx, reused| {
+                    SignatureScriptBuilder::Mixed(vec![create_ecdsa_signature(tx, reused), create_schnorr_signature(tx, reused)])
+                }),
+                expected_sig_ops: 2,
+                sig_op_limit: 2,
+                should_pass: true,
+            },
+            // 2-of-3 MultiSig test case
+            TestCase {
+                name: "2-of-3 MultiSig",
+                script_builder: Box::new(move |sb| {
+                    sb.add_i64(2)?
+                        .add_data(&keypair.x_only_public_key().0.serialize())?
+                        .add_data(&keypair.x_only_public_key().0.serialize())?
+                        .add_data(&keypair.x_only_public_key().0.serialize())?
+                        .add_i64(3)?
+                        .add_op(OpCheckMultiSig)
+                }),
+                sig_builder: Box::new(move |tx, reused| {
+                    let sig = create_schnorr_signature(tx, reused);
+                    SignatureScriptBuilder::Multi(vec![sig.clone(), sig])
+                }),
+                expected_sig_ops: 2,
+                sig_op_limit: 2,
+                should_pass: true,
+            },
+            TestCase {
+                name: "Mixed Schnorr and ECDSA",
```

Comment:
I think it's better not to repeat test names. In this case, adding " - Above limit" or " - Exceeds limit" would be helpful (same for similar tests)

Reply:
f032de3a7ae364278470a15aec4ebdd64c47e9b4